ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/tasks/GetPivotBlockHeaderFromPeerTask.java

```diff
@@ -0,0 +1,93 @@
+import tech.pegasys.pantheon.ethereum.eth.manager.EthContext;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthPeer;
+import tech.pegasys.pantheon.ethereum.eth.manager.exceptions.NoAvailablePeersException;
+import tech.pegasys.pantheon.ethereum.eth.manager.exceptions.PeerBreachedProtocolException;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.metrics.LabelledMetric;
+import tech.pegasys.pantheon.metrics.OperationTimer;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class GetPivotBlockHeaderFromPeerTask extends AbstractRetryingPeerTask<List<BlockHeader>> {
+  private static final Logger LOG = LogManager.getLogger();
+  private final ProtocolSchedule<?> protocolSchedule;
+  private final EthContext ethContext;
+  private final LabelledMetric<OperationTimer> ethTasksTimer;
+  private final Optional<EthPeer> peer;
+  private final long pivotBlockNumber;
+
+  private GetPivotBlockHeaderFromPeerTask(
+      final ProtocolSchedule<?> protocolSchedule,
+      final EthContext ethContext,
+      final LabelledMetric<OperationTimer> ethTasksTimer,
+      final Optional<EthPeer> peer,
+      final long pivotBlockNumber,
+      final int maxRetries) {
+    super(ethContext, maxRetries, ethTasksTimer);
+    this.protocolSchedule = protocolSchedule;
+    this.ethContext = ethContext;
+    this.ethTasksTimer = ethTasksTimer;
+    this.peer = peer;
+    this.pivotBlockNumber = pivotBlockNumber;
+  }
+
+  public static GetPivotBlockHeaderFromPeerTask forPivotBlock(
+      final ProtocolSchedule<?> protocolSchedule,
+      final EthContext ethContext,
+      final LabelledMetric<OperationTimer> ethTasksTimer,
+      final Optional<EthPeer> peer,
+      final long pivotBlockNumber,
+      final int maxRetries) {
+    return new GetPivotBlockHeaderFromPeerTask(
+        protocolSchedule, ethContext, ethTasksTimer, peer, pivotBlockNumber, maxRetries);
+  }
+
+  @Override
+  protected CompletableFuture<List<BlockHeader>> executePeerTask() {
+    final AbstractGetHeadersFromPeerTask getHeadersTask =
+        GetHeadersFromPeerByNumberTask.forSingleNumber(
+            protocolSchedule, ethContext, pivotBlockNumber, ethTasksTimer);
+    peer.ifPresent(getHeadersTask::assignPeer);
+    return executeSubTask(getHeadersTask::run)
+        .thenApply(
+            peerResult -> {
+              if (!peerResult.getResult().isEmpty()) {
+                result.get().complete(peerResult.getResult());
+              }
+              return peerResult.getResult();
+            });
+  }
+
+  public CompletableFuture<BlockHeader> getPivotBlockHeader() {
+    return run().thenApply(singletonList -> singletonList.get(0));
+  }
+
+  @Override
+  protected boolean isRetryableError(final Throwable error) {
```

Comment:
I think we should probably fold the implementation of this method into `AbstractRetryingPeerTask` along with the [`assignPeer()`](https://github.com/PegaSysEng/pantheon/blob/master/ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/tasks/CompleteBlocksTask.java#L109-L112) functionality in `CompleteBlocksTask` and that implemenation of [`isRetryableError`](https://github.com/PegaSysEng/pantheon/blob/master/ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/tasks/CompleteBlocksTask.java#L100-L107).  The difference in that implementation is that if there is an assigned peer, and that peer is missing or errors out there's no reason to retry.

Reply:
Done.